### PR TITLE
Implement dpik for automatic bandwidth selection

### PR DESCRIPTION
**Pull request by tvatter**

### CODE REVIEW

File: R/kde1d.R [hunk 1]
```diff
@@ -12,7 +12,11 @@
 #' @param mult positive bandwidth multiplier; the actual bandwidth used is
 #'   \eqn{bw*mult}.
 #' @param bw bandwidth parameter; has to be a positive number or `NA`; the
-#'   latter calls [`KernSmooth::dpik()`] for automatic selection (default).
+#'   latter uses the direct plug-in methodology, where unknown functionals that
```
Comment (tnagler) [inline R/kde1d.R]: I would simply write 
> uses the direct plug-in methodology of Sheather and Jones (1991).

and add the reference
> Sheather, S. J. and Jones, M. C. (1991). A reliable data-based bandwidth selection method for kernel density estimation. Journal of the Royal Statistical Society, Series B, 53, 683–690.

File: src/wrappers.cpp [hunk 2]
```diff
@@ -93,3 +94,40 @@ Eigen::VectorXd qkde1d_cpp(const Eigen::VectorXd& x,
...
     return q;
 }
+
+// Bandwidth for Kernel Density Estimation
+//' @param x vector of observations
+//' @param grid_size number of equally-spaced points over which binning is
+//' performed to obtain kernel functional approximation
+//' @return the selected bandwidth
+//' @noRd
+// [[Rcpp::export]]
```
Comment (tnagler) [inline src/wrappers.cpp:104]: Not sure this qualifies as wrapper ;)

File: R/kde1d.R (hunk 1 above)
Comment (tvatter) [reply R/kde1d.R]: Wilco, I'll modify when you decide for the other comment

File: src/wrappers.cpp (hunk 2 above)
Comment (tvatter) [reply src/wrappers.cpp:104]: Yeah, I wondered after looking at `qkde1d_cpp` what would qualify as a wrapper.

Initially, I add a `dpik` function in `dpik.hpp` and then a `dpik_cpp` function that simply called `dpik`, but it felt really useless.

Up to you.

Comment (tnagler) [reply src/wrappers.cpp:104]: OK, I see your point. The actual wrapper would be `select_bw()`. How about we port this to to cpp as `select_bw_cpp()` and move `dpik` to `dpik.hpp`?

File: inst/include/dpik.hpp [hunk 3]
```diff
@@ -80,3 +80,45 @@ double bkfe(const Eigen::VectorXd& x,
...
+
+        double alpha = std::pow(std::pow(2.0, 11.0 / 2.0)/(7.0 * n), 1.0 / 9.0);
+        double psi6hat = bkfe(x2, 6, alpha, sa, sb);
+        alpha = std::pow(-3.0 * std::sqrt(2.0 / M_PI)/(psi6hat * n), 1.0 / 7.0);
+        double psi4hat = bkfe(x2, 4, alpha, sa, sb);
+
+        double del0 = 1.0 / std::pow(4 * M_PI, 1/10.0);
+        bw = scale * del0 * std::pow(1.0 / (psi4hat * n), 1.0 / 5.0);
+    } catch (...) {
+        bw = 4 * 1.06 *scale * std::pow(1 / n, 1.0 / 5.0);
```
Comment (tnagler) [inline inst/include/dpik.hpp]: 4.0 and 1.0

**Pull request merged**